src/beast/beast/hash/hash_append.h

```diff
@@ -173,454 +213,256 @@ static_assert (is_contiguously_hashable <
-    T1, T2, T3, T4, T5, T6, T7, T8, T9> const& t) noexcept;
-#endif
+hash_append(Hasher& h, std::unordered_set<Key, Hash, Pred, Alloc> const& s);
 
-#if BEAST_VARIADIC_MAX >= 10
-template <class Hasher, class T1, class T2, class T3, class T4, class T5,
-                        class T6, class T7, class T8, class T9, class T10>
-inline
+template <class Hasher, class T0, class T1, class ...T>
 void
-hash_append (Hasher& h, std::tuple <
-    T1, T2, T3, T4, T5, T6, T7, T8, T9, T10> const& t) noexcept;
-#endif
-
-#endif // BEAST_NO_TUPLE_VARIADICS
-
-//------------------------------------------------------------------------------
+hash_append (Hasher& h, T0 const& t0, T1 const& t1, T const& ...t) noexcept;
 
-namespace detail {
+// c-array
 
-template <class Hasher, class T>
-inline
-int
-hash_one (Hasher& h, T const& t) noexcept
+template <class Hasher, class T, std::size_t N>
+std::enable_if_t
+<
+    !is_contiguously_hashable<T, Hasher>::value
+>
+hash_append(Hasher& h, T (&a)[N]) noexcept
 {
-    hash_append (h, t);
-    return 0;
+    for (auto const& t : a)
+        hash_append(h, t);
 }
 
-} // detail
-
-//------------------------------------------------------------------------------
-
-// std::tuple
+// basic_string
 
-template <class Hasher>
+template <class Hasher, class CharT, class Traits, class Alloc>
 inline
-void
-hash_append (Hasher& h, std::tuple<> const& t) noexcept
+std::enable_if_t
+<
+    !is_contiguously_hashable<CharT, Hasher>::value 
+>
+hash_append(Hasher& h, std::basic_string<CharT, Traits, Alloc> const& s) noexcept
 {
-    hash_append (h, nullptr);
+    for (auto c : s)
+        hash_append(h, c);
+    hash_append(h, s.size());
 }
 
-#if BEAST_NO_TUPLE_VARIADICS
-
-#if BEAST_VARIADIC_MAX >= 1
-template <class Hasher, class T1>
+template <class Hasher, class CharT, class Traits, class Alloc>
 inline
-void
-hash_append (Hasher& h, std::tuple <T1> const& t) noexcept
+std::enable_if_t
+<
+    is_contiguously_hashable<CharT, Hasher>::value 
+>
+hash_append(Hasher& h, std::basic_string<CharT, Traits, Alloc> const& s) noexcept
 {
-    hash_append (h, std::get<0>(t));
+    h(s.data(), s.size()*sizeof(CharT));
+    hash_append(h, s.size());
 }
-#endif
 
-#if BEAST_VARIADIC_MAX >= 2
-template <class Hasher, class T1, class T2>
-inline
-void
-hash_append (Hasher& h, std::tuple <T1, T2> const& t) noexcept
-{
-    hash_append (h, std::get<0>(t));
-    hash_append (h, std::get<1>(t));
-}
-#endif
+// pair
 
-#if BEAST_VARIADIC_MAX >= 3
-template <class Hasher, class T1, class T2, class T3>
+template <class Hasher, class T, class U>
 inline
-void
-hash_append (Hasher& h, std::tuple <T1, T2, T3> const& t) noexcept
+std::enable_if_t
+<
+    !is_contiguously_hashable<std::pair<T, U>, Hasher>::value
+>
+hash_append (Hasher& h, std::pair<T, U> const& p) noexcept
 {
-    hash_append (h, std::get<0>(t));
-    hash_append (h, std::get<1>(t));
-    hash_append (h, std::get<2>(t));
+    hash_append (h, p.first, p.second);
```

Comment:
If I'm reading this correctly, the fact that the data is in a `std::pair` does not contribute to the hash. (Unlike a `vector`, for example, which includes the size of the vector.) Is this intentional?

I think it is, but I want to confirm.